### ISSUE TITLE
Add rule `no-unused-props`

### ISSUE THREAD

**Issue by oekazuma** [sveltejs/eslint-plugin-svelte#1028]
### Motivation

This rule aims to ensure consistency between the defined `Props` type and the actual use of `$props()`.
 This prevents mistakes due to overlooking properties and improves code readability and maintainability. 
Enforcing this rule helps developers understand and reduces the risk of potential bugs.

### Description

Issue a warning if the type of the defined `Props` does not match the variable destructuring assignment in `$props()`.

### Examples

```svelte
<script lang="ts">
  /* ✓ GOOD */

  interface Props {
    a: string;
    b: number;
  }

  let { a, b }: Props = $props();

  /* ✗ BAD */

  interface Props {
    a: string;
    b: number;
  }

  let { a }: Props = $props();
</script>
```

**Comment by ota-meshi** [comment]
Thank you for proposing the rule!
That rule sounds good to me! 


If possible, it would be even better if it could check the usage of unnamed props as well.

```svelte
<script lang="ts">
	interface Props {
		a: number;
		[key: string]: unknown; // unused
	}

	let {a}: Props = $props();
        // We can guess that the following would be correct:
	// let {a, ...other}: Props = $props();
</script>
```

And it would be even better if the rule could check member access as well.

```svelte
<script lang="ts">
  interface Props {
    a: string;
    b: number; // unused
  }

  let props: Props = $props();
</script>

{props.a}
```

```svelte
<script lang="ts">
  interface Props {
    a: string;
    b: number;
    c: number; // unused
  }

  let { a, ...otherProps }: Props = $props();
</script>

{otherProps.b}
```

I think the rule name might be better, something like `no-unused-props`.
What do you think?

**Comment by oekazuma** [comment]
Thanks for the good additional suggestions!

I agree with the rule name, `no-unused-props`.

**Comment by shumadrid** [comment]
Did the usual updating of dependencies, ran eslint, got this output produced by this rule, is this intended?:

```
 27:9  error  'toString' in 'currentDayCommitHash' is an unused property           svelte/no-unused-props
  27:9  error  'charAt' in 'currentDayCommitHash' is an unused property             svelte/no-unused-props
  27:9  error  'charCodeAt' in 'currentDayCommitHash' is an unused property         svelte/no-unused-props
  27:9  error  'concat' in 'currentDayCommitHash' is an unused property             svelte/no-unused-props
  27:9  error  'indexOf' in 'currentDayCommitHash' is an unused property            svelte/no-unused-props
  27:9  error  'lastIndexOf' in 'currentDayCommitHash' is an unused property        svelte/no-unused-props
  27:9  error  'localeCompare' in 'currentDayCommitHash' is an unused property      svelte/no-unused-props
  27:9  error  'match' in 'currentDayCommitHash' is an unused property              svelte/no-unused-props
  27:9  error  'replace' in 'currentDayCommitHash' is an unused property            svelte/no-unused-props
  27:9  error  'search' in 'currentDayCommitHash' is an unused property             svelte/no-unused-props
  27:9  error  'slice' in 'currentDayCommitHash' is an unused property              svelte/no-unused-props
  27:9  error  'split' in 'currentDayCommitHash' is an unused property              svelte/no-unused-props
  27:9  error  'substring' in 'currentDayCommitHash' is an unused property          svelte/no-unused-props
  27:9  error  'toLowerCase' in 'currentDayCommitHash' is an unused property        svelte/no-unused-props
  27:9  error  'toLocaleLowerCase' in 'currentDayCommitHash' is an unused property  svelte/no-unused-props
  27:9  error  'toUpperCase' in 'currentDayCommitHash' is an unused property        svelte/no-unused-props
  27:9  error  'toLocaleUpperCase' in 'currentDayCommitHash' is an unused property  svelte/no-unused-props
  27:9  error  'trim' in 'currentDayCommitHash' is an unused property               svelte/no-unused-props
  27:9  error  'length' in 'currentDayCommitHash' is an unused property             svelte/no-unused-props
  ...
```

currentDayCommitHash is just a regular string passed as a prop.

**Comment by baseballyama** [comment]
@shumadrid Can you share us code example?

**Comment by AndersRobstad** [comment]
@baseballyama Here is a minimal repro. It reports object-props as unused even when they have all their attributes referenced. It should probably allow for objects as long as at least one of its attributes is used inside the file

<img width="795" alt="Image" src="https://github.com/user-attachments/assets/6b2173f2-3995-416e-815b-1ee8d747afcb" />

**Comment by baseballyama** [comment]
@shumadrid @AndersRobstad We’ve released v3.2.1. Could you please verify if this issue has been resolved? As this rule is relatively complex, there might still be some false positives.

**Comment by AndersRobstad** [comment]
@baseballyama We're still seeing this reported as an error, even though the type prop is being referenced.

In our codebase (_and I assume many others as well_), we often pass objects as props to components instead of just primitive values. This rule feels overly strict in our case since it requires every attribute of an object prop to be explicitly referenced for it to be considered valid.This makes it impractical for files where complex objects are passed as props, as we'd need to reference all of their attributes just to satisfy the rule.

<img width="684" alt="Image" src="https://github.com/user-attachments/assets/f6b3b626-eb4a-4cd7-891f-d76ba19372d4" />

**Comment by AndersRobstad** [comment]
> [@baseballyama](https://github.com/baseballyama) We're still seeing this reported as an error, even though the type prop is being referenced.
> 
> In our codebase (_and I assume many others as well_), we often pass objects as props to components instead of just primitive values. This rule feels overly strict in our case since it requires every attribute of an object prop to be explicitly referenced for it to be considered valid.This makes it impractical for files where complex objects are passed as props, as we'd need to reference all of their attributes just to satisfy the rule.
> 
> <img alt="Image" width="684" src="https://private-user-images.githubusercontent.com/54552345/423414982-f6b3b626-eb4a-4cd7-891f-d76ba19372d4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIyMDU1MjQsIm5iZiI6MTc0MjIwNTIyNCwicGF0aCI6Ii81NDU1MjM0NS80MjM0MTQ5ODItZjZiM2I2MjYtZWI0YS00Y2Q3LTg5MWYtZDc2YmExOTM3MmQ0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzE3VDA5NTM0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQyNzYxN2NkYmQxZGI2MDY3M2JkYTc0MmI2YjljNmQzY2UzOGNlZTk1ZDRmZmM1ZDg3ZDU1ZjVjNDU3MzExNmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.tSQ7w_ikwEdt5Ku9wlz6VLcrXf_UTBCCHg-pph3Yy0o">

Seems to work fine for classes now though! They correctly only report as unused if they are never referenced at all

**Comment by baseballyama** [comment]
@AndersRobstad

By default, the `checkImportedTypes` option is set to `false`. This means the rule does not report unused properties for imported types (where the component itself doesn’t own the type definition). On the other hand, types declared within a Svelte component are owned by that component, and thus unused properties should ideally be removable.

Could you please tell me again about the practical issues you’re encountering with your specific use case?

Additionally, we can use the `ignorePatterns` option to exclude specific patterns. This allows you to balance the rule with practical needs in real-world projects.
https://sveltejs.github.io/eslint-plugin-svelte/rules/no-unused-props/

**Comment by AndersRobstad** [comment]
@baseballyama Then the `checkImportedTypes` option does not seem to work correctly. In the example below the svelte-component imports the `TaskViewModel`-type, but it still reports errors due to some of its attributes not being used

<img width="836" alt="Image" src="https://github.com/user-attachments/assets/fbd074c9-26c5-4eb7-9e4c-3c2892f42dab" />

**Comment by baseballyama** [comment]
@AndersRobstad Thank you! But this is strange.🤔
I’ll investigate why this is happening.

**Comment by shumadrid** [comment]
My issue is still here on v3.2.1. This is how i define the string, this might be irrelevant because i get the errors for all primitive types passed as props. The string that's passed is a property in an abstract class:

```ts
interface Properties {
    currentDayCommitHash: string;
    // ...
  }

  const {
    currentDayCommitHash,
    // ...
  }: Properties = $props();
```

My eslint.config.mjs:

```js
export default typescriptEslint.config(
  eslint.configs.recommended,
  ...typescriptEslint.configs.recommendedTypeChecked,
  ...svelte.configs['flat/recommended'],
  ...svelte.configs['flat/prettier'],
  eslintPluginUnicorn.configs.recommended,
  devUtilsConfigs,
  eslintConfigPrettier,
  {
    files: ['**/*.{js,ts}'],
    languageOptions: {
      parser: typescriptEslint.parser,
      parserOptions: {
        projectService: true
      }
    }
  },
  {
    files: ['**/*.svelte'],
    languageOptions: {
      parser: svelteParser,
      parserOptions: {
        parser: typescriptEslint.parser,
        extraFileExtensions: ['.svelte'],
        projectService: true
      }
    },
    rules: {
      'svelte/require-each-key': 'off'
      // 'svelte/no-unused-props': 'off'
    }
  }
// ...
}
```

**Comment by AndersRobstad** [comment]
There still seems to be some problems with the reporting of unused props, created a new issue [here](https://github.com/sveltejs/eslint-plugin-svelte/issues/1145)
